usr/src/lib/libc/stdio/printf.3
.\" Copyright (c) 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Chris Torek.
.\"
.\" %sccs.include.redist.man%
.\"
.\"	@(#)printf.3	6.10 (Berkeley) %G%
.\"
.TH PRINTF 3 ""
.UC 7
.SH NAME
printf, fprintf, sprintf, snprintf, vprintf, vfprintf,
vsprintf, vsnprintf \- formatted output conversion
.SH SYNOPSIS
.B #include <stdio.h>
.PP
.B printf(format
[ , arg ... ]
.B )
.br
.B char *format;
.PP
.B fprintf(stream, format
[ , arg ... ]
.B )
.br
.SM
.B FILE
.B *stream;
.br
.B char *format;
.PP
.B sprintf(str, format
[ , arg ... ]
.B )
.br
.B char *str, *format;
.PP
.B snprintf(str, size, format
[ , arg ... ]
.B )
.br
.B char *str;
.br
.B size_t size;
.br
.B char *format;
.\" .PP
.\" .B char *smprintf(format
.\" [ , arg ... ]
.\" .B )
.PP
.B #include <varargs.h>
.PP
.B vprintf(format, ap)
.br
.B char *format;
.br
.B va_list ap;
.PP
.B vfprintf(stream, format, ap)
.br
.SM
.B FILE
.B *stream;
.br
.B char *format;
.br
.B va_list ap;
.PP
.B vsprintf(str, format, ap)
.br
.B char *str, *format;
.br
.B va_list ap;
.PP
.B vsnprintf(str, size, format, ap)
.br
.B char *str;
.br
.B size_t size;
.br
.B char *format;
.br
.B va_list ap;
.\" .PP
.\" .B char *vsmprintf(format, ap)
.\" .br
.\" .B char *format;
.\" .br
.\" .B va_list ap;
.SH DESCRIPTION
The
.I printf
family of functions produces output according to a
.I format
as described below.
.I Printf
and
.I vprintf
write the standard output stream
.BR stdout ;
.I fprintf
and
.I vfprintf
write the named output
.IR stream ;
.IR sprintf ,
.IR snprintf ,
.IR vsprintf ,
and
.I vsnprintf
write to the character string
.IR str .
.\" .IR str ;
.\" and
.\" .I smprintf
.\" and
.\" .I vsmprintf
.\" dynamically allocate a new string with
.\" .IR malloc .
Each function converts, formats, and prints the arguments following the
.I format
(or accessed via the variable-length argument facilities of
.IR varargs (3))
under control of the
.I format
argument.
.\" Except for
.\" .I smprintf
.\" and
.\" .IR vsmprintf ,
.\" all of these functions return
Each function returns
the number of characters printed
(not including the trailing `\e0' used to end output to strings).
.\" .I Smprintf
.\" and
.\" .I vsmprintf
.\" return a pointer to a string of an appropriate length;
.\" this pointer should be passed to
.\" .I free
.\" to release the associated storage
.\" when it is no longer needed.
.\" If sufficient space is not avaliable,
.\" .I smprintf
.\" and
.\" .I vsmprintf
.\" will return
.\" .SM
.\" .BR NULL .
.I Snprintf
and
.I vsnprintf
will write at most
.IR size \-1
of the characters printed into the output string
(the
.IR size 'th
character then gets the terminating `\e0');
if the return value is greater than or equal to the
.I size
argument, the string was too short
and some of the printed characters were discarded.
.I Sprintf
and
.I vsprintf
effectively assume an infinte
.IR size .
.PP
The format is a string comprised of two types of objects:
plain characters, which are simply copied to the output stream,
and conversion specifications,
each of which causes conversion and printing of the next successive
.IR arg .
These arguments must correspond properly (after type promotion)
with the conversion specifier.
Each conversion is introduced by the
.B %
(percent sign) character.
The rest of the conversion includes, in the following order,
.TP 4
.B \(bu
Zero or more of the following flags:
.RS
.TP 4
.B \(bu
a `#' character
specifying that the value should be converted to an ``alternate form''.
For 
.BR c ,
.BR d ,
.BR i ,
.BR n ,
.BR p ,
.BR s ,
and
.BR u ,
conversions, this option has no effect.
For 
.B o
conversions, the precision of the number is increased to force the first
character of the output string to a zero (except if a zero value is printed
with an explicit precision of zero).
For
.B x
and
.B X
conversions, a non-zero result has the string
.B 0x
(or
.B 0X
for
.B X
conversions) prepended to it.
For
.BR e ,
.BR E ,
.BR f ,
.BR g ,
and
.BR G ,
conversions, the result will always contain a decimal point, even if no
digits follow it (normally, a decimal point appears in the results of
those conversions only if a digit follows).
For
.B g
and
.B G
conversions, trailing zeros are not removed from the result as they
would otherwise be.
.TP 4
.B \(bu
A zero `0' character specifying zero padding.
For all conversions except
.BR n ,
the converted value is padded on the left with zeros rather than blanks.
If a precision is given with a numeric conversion
.RB ( d ,
.BR i ,
.BR o ,
.BR u ,
.BR i ,
.BR x ,
and
.BR X ),
the `0' flag is ignored.
.TP 4
.B \(bu
A minus sign (`-') specifying left adjustment of the converted value
in the indicated field.
Except for
.B n
conversions, the converted value is padded on the right with blanks,
rather than on the left with blanks or zeros.
A `-' overrides a `0' if both are given.
.TP 4
.B \(bu
A space, specifying that a blank should be left before a positive number
produced by a signed conversion
.RB ( d ,
.BR e ,
.BR E ,
.BR f ,
.BR g ,
.BR G ,
or
.BR i ).
.TP 4
.B \(bu
a `+' character specifying that a sign always be placed before a
number produced by a signed conversion.
A `+' overrides a space if both are used.
.RE
.TP 4
.B \(bu
An optional digit string specifying a field width.
If the converted value has fewer characters than the field width, it will
be padded on the left (or right, if the left-adjustment flag is used) to
make up the field width.
.TP 4
.B \(bu
An optional precision, in the form of a period (`.') followed by an
optional digit string.  If the digit string is omitted, the precision
is taken as zero.  This gives the minimum number of digits to appear for
.BR d ,
.BR i ,
.BR o ,
.BR u ,
.BR x ,
and
.B X
conversions, the number of digits to appear after the decimal point for
.BR e ,
.BR E ,
and
.B f
conversions, the maximum number of significant digits for
.B g
and
.B G
conversions, or the maximum number of characters to be printed from a
string for
.B s
conversions.
.TP 4
.B \(bu
The character
.BR h ,
specifying that a following
.BR d ,
.BR i ,
.BR o ,
.BR u ,
.BR x ,
or
.B X
conversion corresponds to a
.B "short int"
or
.B "unsigned short int"
argument, or that a following
.B n
conversion corresponds to a pointer to a
.B "short int"
argument.
.TP 4
.B \(bu
the character
.B l
(ell) specifying that a following
.BR d ,
.BR i ,
.BR o ,
.BR u ,
.BR x ,
or
.B X
conversion corresponds to a
.B "long int"
or
.B "unsigned long int"
argument, or that a following
.B n
conversion corresponds to a pointer to a
.B "long int"
argument.
.TP 4
.B \(bu
The character
.B L
specifying that a following
.BR e ,
.BR E ,
.BR f ,
.BR g ,
or
.B G
conversion corresponds to a
.B "long double"
argument (but note that long double values are not currently supported
by the \s-2VAX\s0 and Tahoe compilers).
.TP 4
.B \(bu
A character which indicates the type of conversion to be applied.
.PP
A field width or precision may be an asterisk (`*') instead of a
digit string.
In this case an
.B int
argument supplies the value.
A negative field width is treated as a left adjustment flag followed by a
positive field width; a negative precision is treated as though it were
missing.
.PP
The conversion characters and their meanings are:
.TP 4
.B diouxX
The
.B int
(or appropriate variant) argument is converted to signed decimal
.RB ( d " and " i ),
unsigned octal
.RB ( o ),
unsigned decimal
.RB ( u ),
or unsigned hexadecimal
.RB ( x " and " X )
notation respectively.  The letters
.B abcdef
are used for
.B x
conversions; the letters
.B ABCDEF
are used for
.B X
conversions.
The precision, if any, gives the minimum number of digits that must
appear; if the converted value requires fewer digits, it is padded on
the left with zeros.
.TP 4
.B DOU
The
.B "long int"
argument is converted to signed decimal, unsigned octal, or unsigned
decimal, as if the format had been
.BR ld ,
.BR lo ,
or
.B lu
respectively.
These conversion characters are deprecated, and will eventually disappear.
.TP 4
.B eE
The
.B double
argument is rounded and converted in the style
`[\fB\-\fR]d\fB.\fRddd\fBe\fR\(+-dd' where there is one digit before the
decimal point and the number after is equal to the precision specification
for the argument.
If the precision is missing, 6 digits are given; if the precision is
explicitly zero, no decimal point appears.
An
.B E
conversion uses the letter
.B E
(rather than
.BR e )
to introduce the exponent.
The exponent always contains at least two digits; if the value is zero,
the exponent is 00.
.TP 4
.B f
The
.B double
argument is rounded and converted to decimal notation in the style
`[\fB\-\fR]ddd.ddd' where the number of digits after the decimal point
is equal to the precision.
If the precision is missing, 6 digits are given; if the precision is
explicitly 0, no digits and no decimal point are printed.
If a decimal point appears, at least one digit appears before it.
.TP 4
.B g
The
.B double
argument is printed in style
.B f
or
.B e
(or
.B E
for
.B G
conversions).
The precision specifies the number of significant digits.
If the precision is missing, 6 digits are given; if the precision is zero,
it is treated as 1.
Style
.B e
is used if the exponent from its conversion is less than -4 or greater than
or equal to the precision.
Trailing zeros are removed from the fractional part of the result; a
decimal point appears only if it is followed by at least one digit.
.TP 4
.B c
The
.B int
argument is converted to an
.B "unsigned char",
and the resulting character is printed.
.TP 4
.B s
The
.B "char *"
argument is taken to be a string (character pointer).
Characters from the string are printed until a null character is reached,
or until the number of characters indicated by the precision have been
printed, whichever occurs first; if a precision is given, no null character
need be present.
.TP 4
.B p
The
.B "void *"
pointer argument is printed in hexadecimal (as if by `%#x' or `%#lx').
.TP 4
.B n
The number of characters written so far is stored into the
integer indicated by the
.B "int *"
(or variant) pointer argument.
No argument is converted.
.TP 4
.B %
Prints a `%'; no argument is converted.
.PP
In no case does a non-existent or small field width cause truncation of
a field; if the result of a conversion is wider than the field width, the
field is expanded to contain it.
Similarly, padding takes place only if the specified field width exceeds
the actual width.
.PP
.SH EXAMPLES
.br
To print a date and time in the form `Sunday, July 3, 10:02',
where
.I weekday
and
.I month
are pointers to null-terminated strings:
.RS
.HP
.nh
printf("%s, %s %d, %02d:%.2d", weekday, month, day, hour, min);
.RE
.hy
.PP
To print
.if n pi
.if t \(*p
to 5 decimals:
.IP
printf("pi = %.5f", 4*atan(1.0));
.PP
To allocate a 128 byte string and print into it:
.RS
.nf
.ta 1i 2i
.sp
#include <stdio.h>
#include <varargs.h>
char *newfmt(va_alist)
	va_dcl
{
	char *p, *malloc(), *fmt;
	va_list ap;
	if ((p = malloc(128)) == NULL)
		return (NULL);
	va_start(ap);
	fmt = va_arg(ap, char *);
	(void) vsprintf(p, fmt, ap);
	va_end(ap);
	return (p);
}
.RE
.fi
.SH "SEE ALSO"
printf(1), scanf(3)
.SH BUGS
The conversion formats `%D', `%O', and `%U' are not standard and
are provided only for backward compatibility.
The effect of padding the `%p' format with zeros (either by the `0'
flag or by specifying a precision), and the benign effect (i.e., none)
of the `#' flag on `%n' and `%p' conversions, as well as other
nonsensical combinations such as `%Ld', are not standard; such combinations
should be avoided.
.PP
Because
.I sprintf
and
.I vsprintf
assume an infinitely long string,
callers must be careful not to overflow the actual space;
this is often impossible to assure.
For safety, programmers should use the
.I snprintf
interface instead.
Unfortunately, this interface is not portable.


usr/src/usr.bin/ex/printf.c
/*-
 * Copyright (c) 1980 The Regents of the University of California.
 * All rights reserved.
 *
 * %sccs.include.proprietary.c%
 */

#ifndef lint
static char sccsid[] = "@(#)printf.c	7.6 (Berkeley) %G%";
#endif /* not lint */

#ifndef lint
/* The pwb version this is based on */
static char *printf_id = "@(#) printf.c:2.2 6/5/79";
#endif /* not lint */

#include <varargs.h>
 
/*
 * This version of printf is compatible with the Version 7 C
 * printf. The differences are only minor except that this
 * printf assumes it is to print through putchar. Version 7
 * printf is more general (and is much larger) and includes
 * provisions for floating point.
 */

#define MAXOCT	11	/* Maximum octal digits in a long */
#define MAXINT	32767	/* largest normal length positive integer */
#define BIG	1000000000  /* largest power of 10 less than an unsigned long */
#define MAXDIGS	10	/* number of digits in BIG */

static int width, sign, fill;

char *_p_dconv();

/* VARARGS */
ex_printf(va_alist)
	va_dcl
{
	va_list ap;
	register char *fmt;
	char fcode;
	int prec;
	int length,mask1,nbits,n;
	long int mask2, num;
	register char *bptr;
	char *ptr;
	char buf[134];

	va_start(ap);
	fmt = va_arg(ap,char *);
	for (;;) {
		/* process format string first */
		while ((fcode = *fmt++)!='%') {
			/* ordinary (non-%) character */
			if (fcode=='\0')
				return;
			ex_putchar(fcode);
		}
		/* length modifier: -1 for h, 1 for l, 0 for none */
		length = 0;
		/* check for a leading - sign */
		sign = 0;
		if (*fmt == '-') {
			sign++;
			fmt++;
		}
		/* a '0' may follow the - sign */
		/* this is the requested fill character */
		fill = 1;
		if (*fmt == '0') {
			fill--;
			fmt++;
		}
		
		/* Now comes a digit string which may be a '*' */
		if (*fmt == '*') {
			width = va_arg(ap, int);
			if (width < 0) {
				width = -width;
				sign = !sign;
			}
			fmt++;
		}
		else {
			width = 0;
			while (*fmt>='0' && *fmt<='9')
				width = width * 10 + (*fmt++ - '0');
		}
		
		/* maybe a decimal point followed by more digits (or '*') */
		if (*fmt=='.') {
			if (*++fmt == '*') {
				prec = va_arg(ap, int);
				fmt++;
			}
			else {
				prec = 0;
				while (*fmt>='0' && *fmt<='9')
					prec = prec * 10 + (*fmt++ - '0');
			}
		}
		else
			prec = -1;
		
		/*
		 * At this point, "sign" is nonzero if there was
		 * a sign, "fill" is 0 if there was a leading
		 * zero and 1 otherwise, "width" and "prec"
		 * contain numbers corresponding to the digit
		 * strings before and after the decimal point,
		 * respectively, and "fmt" addresses the next
		 * character after the whole mess. If there was
		 * no decimal point, "prec" will be -1.
		 */
		switch (*fmt) {
			case 'L':
			case 'l':
				length = 2;
				/* no break!! */
			case 'h':
			case 'H':
				length--;
				fmt++;
				break;
		}
		
		/*
		 * At exit from the following switch, we will
		 * emit the characters starting at "bptr" and
		 * ending at "ptr"-1, unless fcode is '\0'.
		 */
		switch (fcode = *fmt++) {
			/* process characters and strings first */
			case 'c':
				buf[0] = va_arg(ap, int);
				ptr = bptr = &buf[0];
				if (buf[0] != '\0')
					ptr++;
				break;
			case 's':
				bptr = va_arg(ap,char *);
				if (bptr==0)
					bptr = "(null pointer)";
				if (prec < 0)
					prec = MAXINT;
				for (n=0; *bptr++ && n < prec; n++) ;
				ptr = --bptr;
				bptr -= n;
				break;
			case 'O':
				length = 1;
				fcode = 'o';
				/* no break */
			case 'o':
			case 'X':
			case 'x':
				if (length > 0)
					num = va_arg(ap,long);
				else
					num = (unsigned)va_arg(ap,int);
				if (fcode=='o') {
					mask1 = 0x7;
					mask2 = 0x1fffffffL;
					nbits = 3;
				}
				else {
					mask1 = 0xf;
					mask2 = 0x0fffffffL;
					nbits = 4;
				}
				n = (num!=0);
				bptr = buf + MAXOCT + 3;
				/* shift and mask for speed */
				do
				    if (((int) num & mask1) < 10)
					*--bptr = ((int) num & mask1) + 060;
				    else
					*--bptr = ((int) num & mask1) + 0127;
				while (num = (num >> nbits) & mask2);
				
				if (fcode=='o') {
					if (n)
						*--bptr = '0';
				}
				else
					if (!sign && fill <= 0) {
						ex_putchar('0');
						ex_putchar(fcode);
						width -= 2;
					}
					else {
						*--bptr = fcode;
						*--bptr = '0';
					}
				ptr = buf + MAXOCT + 3;
				break;
			case 'D':
			case 'U':
			case 'I':
				length = 1;
				fcode = fcode + 'a' - 'A';
				/* no break */
			case 'd':
			case 'i':
			case 'u':
				if (length > 0)
					num = va_arg(ap,long);
				else {
					n = va_arg(ap,int);
					if (fcode=='u')
						num = (unsigned) n;
					else
						num = (long) n;
				}
				if (n = (fcode != 'u' && num < 0))
					num = -num;
				/* now convert to digits */
				bptr = _p_dconv(num, buf);
				if (n)
					*--bptr = '-';
				if (fill == 0)
					fill = -1;
				ptr = buf + MAXDIGS + 1;
				break;
			default:
				/* not a control character, 
				 * print it.
				 */
				ptr = bptr = &fcode;
				ptr++;
				break;
			}
			if (fcode != '\0')
				_p_emit(bptr,ptr);
	}
	va_end(ap);
}

/* _p_dconv converts the unsigned long integer "value" to
 * printable decimal and places it in "buffer", right-justified.
 * The value returned is the address of the first non-zero character,
 * or the address of the last character if all are zero.
 * The result is NOT null terminated, and is MAXDIGS characters long,
 * starting at buffer[1] (to allow for insertion of a sign).
 *
 * This program assumes it is running on 2's complement machine
 * with reasonable overflow treatment.
 */
char *
_p_dconv(value, buffer)
	long value;
	char *buffer;
{
	register char *bp;
	register int svalue;
	int n;
	long lval;
	
	bp = buffer;
	
	/* zero is a special case */
	if (value == 0) {
		bp += MAXDIGS;
		*bp = '0';
		return(bp);
	}
	
	/* develop the leading digit of the value in "n" */
	n = 0;
	while (value < 0) {
		value -= BIG;	/* will eventually underflow */
		n++;
	}
	while ((lval = value - BIG) >= 0) {
		value = lval;
		n++;
	}
	
	/* stash it in buffer[1] to allow for a sign */
	bp[1] = n + '0';
	/*
	 * Now develop the rest of the digits. Since speed counts here,
	 * we do it in two loops. The first gets "value" down until it
	 * is no larger than MAXINT. The second one uses integer divides
	 * rather than long divides to speed it up.
	 */
	bp += MAXDIGS + 1;
	while (value > MAXINT) {
		*--bp = (int)(value % 10) + '0';
		value /= 10;
	}
	
	/* cannot lose precision */
	svalue = value;
	while (svalue > 0) {
		*--bp = (svalue % 10) + '0';
		svalue /= 10;
	}
	
	/* fill in intermediate zeroes if needed */
	if (buffer[1] != '0') {
		while (bp > buffer + 2)
			*--bp = '0';
		--bp;
	}
	return(bp);
}

/*
 * This program sends string "s" to putchar. The character after
 * the end of "s" is given by "send". This allows the size of the
 * field to be computed; it is stored in "alen". "width" contains the
 * user specified length. If width<alen, the width will be taken to
 * be alen. "sign" is zero if the string is to be right-justified
 * in the field, nonzero if it is to be left-justified. "fill" is
 * 0 if the string is to be padded with '0', positive if it is to be
 * padded with ' ', and negative if an initial '-' should appear before
 * any padding in right-justification (to avoid printing "-3" as
 * "000-3" where "-0003" was intended).
 */
_p_emit(s, send)
	register char *s;
	char *send;
{
	char cfill;
	register int alen;
	int npad;
	
	alen = send - s;
	if (alen > width)
		width = alen;
	cfill = fill>0? ' ': '0';
	
	/* we may want to print a leading '-' before anything */
	if (*s == '-' && fill < 0) {
		ex_putchar(*s++);
		alen--;
		width--;
	}
	npad = width - alen;
	
	/* emit any leading pad characters */
	if (!sign)
		while (--npad >= 0)
			ex_putchar(cfill);
			
	/* emit the string itself */
	while (--alen >= 0)
		ex_putchar(*s++);
		
	/* emit trailing pad characters */
	if (sign)
		while (--npad >= 0)
			ex_putchar(cfill);
}
